### PR TITLE
feat: useMediatedState

### PR DESCRIPTION
**Pull request by xobotyi**
This hook will be usefull for forms with inputs and so on, now the function that processes raw input can be defined near the state, not "somwhere there, maybe inlined in jsx props".

### CODE REVIEW

File: src/__stories__/useMediatedState.story.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { storiesOf } from '@storybook/react';
+import * as React from 'react';
+import { useMediatedState } from '../useMediatedState';
+import ShowDocs from './util/ShowDocs';
+
+const InputMediator = s => s.replace(/[\s]+/, ' ');
```
Comment (wardoost) [inline src/__stories__/useMediatedState.story.tsx]: I think the regex needs to be global and only replace double spaces?

```js
const InputMediator = s => s.replace(/\s\s+/g, ' ');
```

Comment (xobotyi) [reply src/__stories__/useMediatedState.story.tsx]: Typo%) recently been writing in php - always forget to puth the `g` flag in JS's regexes..

**Comment by xobotyi** [comment]
@wardoost dont know why, but as for my ears `useMediatedState` sound more suitable here than `useStateMediator`. Second one sound like it wont affect the state, while fors one is for sure will.

File: src/__stories__/useMediatedState.story.tsx (hunk 1 above)
Comment (wardoost) [reply src/__stories__/useMediatedState.story.tsx]: Doesn't the regex have to be `/\s\s+/g` to only replace double spaces?

I would also change `InputMediator` to `inputMediator` as it's the convention in JS.

File: docs/useMediatedState.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# `useMediatedState`
+
+A lot like the standard `useState`, but with mediation process.
+
+## Usage
+```ts
+import * as React from 'react';
+import { useMediatedState } from '../useMediatedState';
+
+const InputMediator = s => s.replace(/[\s]+/, ' ');
```
Comment (wardoost) [inline docs/useMediatedState.md]: Can you update the regex and rename `InputMediator` to `inputMediator`?

Comment (xobotyi) [reply docs/useMediatedState.md]: I've updated %) strange that you've accessed the old one.
About function name i'll forcepush now.

File: src/__stories__/useMediatedState.story.tsx (hunk 1 above)
Comment (xobotyi) [reply src/__stories__/useMediatedState.story.tsx]: There can be not only space sharacters (4ex tab character) so have to replace them too.
Anyways not sure if it is important for the demo.

File: docs/useMediatedState.md (hunk 2 above)
Comment (wardoost) [reply docs/useMediatedState.md]: This comment is in the mardown file of your hook, not the story :)

Comment (xobotyi) [reply docs/useMediatedState.md]: \*facepalm.jpg*

File: src/__stories__/useMediatedState.story.tsx (hunk 1 above)
Comment (wardoost) [reply src/__stories__/useMediatedState.story.tsx]: Doesn't `\s` on its own cover all whitespace characters? I don't think you need the square brackets.

It's not super important for the demo but best to give a common example that is correct/

**Pull request merged**